discovery/endorsement/endorsement.go

```diff
@@ -150,23 +171,51 @@ func (ea *endorsementAnalyzer) computeEndorsementResponse(ctx *context) (*discov
 
 	layouts := computeLayouts(ctx.principalsSets, principalGroups, satGraph)
 	if len(layouts) == 0 {
-		return nil, errors.New("cannot satisfy any principal combination")
+		return nil, errors.New("no peer combination can satisfy the endorsement policy")
 	}
 
 	criteria := &peerMembershipCriteria{
-		possibleLayouts: layouts,
-		satGraph:        satGraph,
-		chanMemberById:  ctx.channelMembersById,
-		idOfMembers:     ctx.identitiesOfMembers,
+		possibleLayouts:  layouts,
+		satGraph:         satGraph,
+		chanMemberById:   ctx.channelMembersById,
+		idOfMembers:      ctx.identitiesOfMembers,
+		chaincodeMapping: ctx.chaincodeMapping,
+	}
+
+	endorsersByGroup := endorsersByGroup(criteria)
+	layouts = filterOutUnsatisfiedLayouts(endorsersByGroup, layouts)
+
+	if len(layouts) == 0 {
+		return nil, errors.New("required chaincodes are not installed on sufficient peers")
 	}
 
 	return &discovery.EndorsementDescriptor{
 		Chaincode:         ctx.chaincode,
 		Layouts:           layouts,
-		EndorsersByGroups: endorsersByGroup(criteria),
+		EndorsersByGroups: endorsersByGroup,
 	}, nil
 }
 
+func filterOutUnsatisfiedLayouts(endorsersByGroup map[string]*discovery.Peers, layouts []*discovery.Layout) []*discovery.Layout {
+	// Iterate once again over all layouts and ensure every layout has enough peers in the EndorsersByGroups
+	// as required by the quantity in the layout.
+	var filteredLayouts []*discovery.Layout
```

Comment:
You could optimise here and pre-create the slice with 0 size and capacity equal to layouts

Reply:
done